### PR TITLE
[direct-core] Introduce offsets on BatchLogReader.

### PR DESCRIPTION
**Pull request by dmvk**

### CODE REVIEW

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java [hunk 1]
```diff
@@ -19,35 +19,54 @@
...
+
+      @Override
+      public Partition getPartition() {
+        return partition;
+      }
+
+      @Override
+      public BatchLogObserver.Offset getOffset() {
+        throw new UnsupportedOperationException(
+            "Unable to calculate offset, because underlying data may have changed.");
```
Comment (je-ik) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java:47]: I'd rephrase this "because the underlying data store is not known to be immutable". That should give better hint what to do, when you encounter this exception.

**Review by je-ik — APPROVED** [review]
LGTM! I'd change the description of exception thrown, but can be merged as is. :+1:

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java (hunk 1 above)
Comment (dmvk) [reply direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java:47]: 👍

**Comment by je-ik** [comment]
Looks like both the coverage and duplication report can be solved by introducing a (tested) wrapper for OnNextContext instead of the anonymous class. Worth it IMHO. Should be simple.
